### PR TITLE
Switch to null for empty metadata

### LINKED ISSUE #352: Switch to default metadata being `null` rather than `{}`?
In tskit, if the metadata schema is JSON, then blank entries are filled with `null` rather than `{}`. In tsinfer (where the JSON format is assumed, and schemas have not yet been implemented), we set blank entries to be `{}` instead. It might help transition to the schema style if we switch to `null` to match tskit. I don't think this will cause any backwards incompatibilities, as `{}` will continue to be valid in old files. It seems like a trivial change to me - can you see any problems with this @jeromekelleher ?

### PR DESCRIPTION
**Pull request by hyanwong**
Fixes #352

### CODE REVIEW

**Comment by hyanwong** [comment]
Weirdly this is complaining about a line that I'm sure is covered, e.g. in

```
nosetests tests/test_inference.py:TestHistoricalSamples
```

- If I raise an error in that part of the loop (`if ind.metadata is None:`) then it triggers in this test. So I have no idea why CI is failing here. If we agree with #352  I think we should just merge this PR, even with CI complaint. it's clearly the correct logic.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -6,12 +6,22 @@
...
   0/1 values as before.
+
 - Times for undated sites now use frequencies (0..1), not as counts (1..num_samples),
   and are now stored as -inf, then calculated on the fly in the variants() iterator.
+
 - The SampleData file no longer accepts the ``inference`` argument to add_site.
   This functionality has been replaced by the ``exclude_positions`` argument
   to the ``infer`` and ``generate_ancestors`` functions.
 
+**New features**:
```
Comment (jeromekelleher) [inline CHANGELOG.rst:17]: That's a breaking change, surely.

File: tests/test_formats.py [hunk 2]
```diff
@@ -478,10 +479,11 @@ def test_from_tree_sequence_omitting_metadata(self):
         self.assertTrue(np.all(sd2.sites_position[:] == sd1.sites_position[:]))
         self.assertTrue(np.all(sd2.sites_genotypes[:] == sd1.sites_genotypes[:]))
         self.assertTrue(np.all(sd2.sites_time[:] == sd1.sites_time[:]))
-        self.assertTrue(np.all(sd2.populations_metadata[:] == {}))
-        self.assertTrue(np.all(sd2.individuals_metadata[:] == {}))
-        self.assertTrue(np.all(sd2.samples_metadata[:] == {}))
-        self.assertTrue(np.all(sd2.sites_metadata[:] == {}))
+        print(sd2.populations_metadata[:])
```
Comment (jeromekelleher) [inline tests/test_formats.py:482]: stray print

**Review by jeromekelleher — COMMENTED** [review]
I'm not sure what we're actually solving with this, and that this isn't a change we're going to end up undoing when we do the metadata schemas properly. 

I think I'd rather keep things as they are, and instead define a schema which permits the empty dict as metadata. That will break less code, IMO

File: CHANGELOG.rst (hunk 1 above)
Comment (hyanwong) [reply CHANGELOG.rst:17]: I don't think it breaks anything, to be honest. We can still read-in sample data files that have `{}` in the metadata file. The reason that I would like to do this is that I want to round-trip data (specifically metadata) from a tree sequence through `.from_tree_sequence()`, via inference, and back to another tree sequence, and this eases that transition, I think.

Comment (jeromekelleher) [reply CHANGELOG.rst:17]: It can break downstream code that depends on the ts having a dictionary in the metadata, not None.

Comment (hyanwong) [reply CHANGELOG.rst:17]: Fair enough - I can add it to breaking changes. But the JSON codec for tree sets allows this (and it's the default), so I think any such code may be problematic anyway (and will stop working when we switch to defaulting to the JSON codec for inferred tree sequences, which I assume we want to do?

Comment (jeromekelleher) [reply CHANGELOG.rst:17]: I just think there's no point in doing this and making the change before we've added the schemas - we're adding extra breakage that we'll have to keep track of and document.

So, add the schemas first. We'll have to do this within a week anyway, so there's no point in putting it off.

**Comment by hyanwong** [comment]
> I think I'd rather keep things as they are, and instead define a schema which permits the empty dict as metadata. That will break less code, IMO

An empty dict *is* permitted as metadata in the tskit JSON schema. It's just that's not the *default*. So we can't round-trip a JSON tree sequence via tsinfer, which is a pain, IMO - you end up changing all the metadata from `null` to `{}`.

You could argue with @benjeffery that the default "empty" metadata field for JSON in tskit should be `{}`, but I assume there's a good reason for choosing `'null'`?

**Comment by jeromekelleher** [comment]
I don't want to get into the details of what the schema should and shouldn't be for tsinfer here. Let's implement the schema first before making backwards incompatible changes.

File: CHANGELOG.rst (hunk 1 above)
Comment (hyanwong) [reply CHANGELOG.rst:17]: Well, if we are doing it this week, then sure!

**Comment by benjeffery** [comment]
> You could argue with @benjeffery that the default "empty" metadata field for JSON in tskit should be `{}`, but I assume there's a good reason for choosing `'null'`?

There is no default metadata for JSON - the default metadata arg to `add_row` has always been `None`.

**Comment by hyanwong** [comment]
> > You could argue with @benjeffery that the default "empty" metadata field for JSON in tskit should be `{}`, but I assume there's a good reason for choosing `'null'`?
> 
> There is no default metadata for JSON - the default metadata arg to `add_row` has always been `None`.

I guess the problem is that when we get a metadata value of `None` passed to tsinfer we actually store `{}`, which is not what tskit does.

**Comment by jeromekelleher** [comment]
Let's just drop this for now @hyanwong - I'll pick up the metadata in tsinfer question once the parsimony stuff is fixed.

**Comment by hyanwong** [comment]
Sure. It turned out to only be a thing for some of the unit tests in https://github.com/tskit-dev/tsinfer/pull/358, which I've just changed with a comment to fix

**Comment by hyanwong** [comment]
Closing in favour of a proper approach to metadata (https://github.com/tskit-dev/tsinfer/issues/272)

**Pull request closed without merging**